### PR TITLE
Add numa-balanced mode to hpx::bind, spread cores over numa domains

### LINKED ISSUE #2899: hpx::bind does not have a mode that balances cores over numa domains

### PR DESCRIPTION
**Pull request by biddisco**
This fixes #2899

### CODE REVIEW

File: src/runtime/threads/policies/parse_affinity_options.cpp [hunk 1]
```diff
@@ -842,6 +843,72 @@ namespace hpx { namespace threads { namespace detail
...
     ///////////////////////////////////////////////////////////////////////////
+    void decode_numabalanced_distribution(hwloc_topology_info& t,
+        std::vector<mask_type>& affinities,
+        std::size_t used_cores, std::size_t max_cores,
+        std::vector<std::size_t>& num_pus, error_code& ec)
+    {
+        std::size_t num_threads = affinities.size();
+        num_pus.resize(num_threads);
+        // numa nodes
+        std::size_t num_numas = (std::max)(std::size_t(1),t.get_number_of_numa_nodes());
```
Comment (sithhell) [inline src/runtime/threads/policies/parse_affinity_options.cpp]: Should we add a fallback to the number of sockets here if the kernel misses NUMA support? We do that already in a couple of places.

File: src/runtime/threads/policies/parse_affinity_options.cpp [hunk 2]
```diff
@@ -842,6 +843,72 @@ namespace hpx { namespace threads { namespace detail
...
+        std::vector<std::size_t> num_cores_numa(num_numas, 0);
+        std::vector<std::size_t> num_threads_numa(num_numas, 0);
+        std::size_t cores_t = 0;
+        for (std::size_t n=0; n<num_numas; ++n) {
+            num_cores_numa[n] = t.get_number_of_numa_node_cores(n);
+            cores_t += num_cores_numa[n];
+        }
+        // how many threads should go on each domain
+        for (std::size_t n=0; n<num_numas; ++n) {
+            num_threads_numa[n] = num_threads*num_cores_numa[n]/cores_t;
```
Comment (sithhell) [inline src/runtime/threads/policies/parse_affinity_options.cpp]: This needs to be rounded up, in the case where, for example, 1 thread is requested, we end up with 0 threads per numa domain.

File: src/runtime/threads/policies/parse_affinity_options.cpp (hunk 1 above)
Comment (biddisco) [reply src/runtime/threads/policies/parse_affinity_options.cpp]: Not really any point. if the kernel lacks numa support then asking for numa-balanced is meaningless anyway. Let it default to 1 domain.

**Comment by biddisco** [comment]
That fixes the 1 thread and any odd number requested  - and also ensures that the rounding up does not oversubscribe the node

File: src/runtime/threads/policies/parse_affinity_options.cpp (hunk 1 above)
Comment (hkaiser) [reply src/runtime/threads/policies/parse_affinity_options.cpp]: What about reporting an error at startup if the user asks for NUMA-domain awareness on a machine which does not have NUMA domains.

**Pull request merged**